google/cloud/pubsub/internal/tracing_batch_callback.cc

```diff
@@ -192,6 +196,58 @@ class TracingBatchCallback : public BatchCallback {
     }
   }
 
+  void StartModackSpan(
+      google::pubsub::v1::ModifyAckDeadlineRequest const& request,
+      std::int64_t nonce) override {
+    namespace sc = opentelemetry::trace::SemanticConventions;
+    using Attributes =
+        std::vector<std::pair<opentelemetry::nostd::string_view,
+                              opentelemetry::common::AttributeValue>>;
+    using Links =
+        std::vector<std::pair<opentelemetry::trace::SpanContext, Attributes>>;
+
+    if (request.ack_ids().empty()) return;
+
+    Links links;
+    std::unique_lock<std::mutex> lk(mu_);
+    for (auto const& ack_id : request.ack_ids()) {
+      auto spans = spans_by_ack_id_.find(ack_id);
+      if (spans != spans_by_ack_id_.end()) {
+        links.emplace_back(std::make_pair(
+            spans->second.subscribe_span->GetContext(), Attributes{}));
+      }
+    }
+    lk.unlock();
+
+    opentelemetry::trace::StartSpanOptions options;
+    options.kind = opentelemetry::trace::SpanKind::kClient;
+    auto span = internal::MakeSpan(
+        subscription_.subscription_id() + " modack",
+        {{sc::kMessagingSystem, "gcp_pubsub"},
+         {sc::kMessagingOperation, "extend"},
+         {sc::kMessagingBatchMessageCount,
+          static_cast<int64_t>(request.ack_ids().size())},
+         {"messaging.gcp_pubsub.message.ack_deadline_seconds",
+          static_cast<int64_t>(request.ack_deadline_seconds())},
+         {sc::kMessagingDestinationName, subscription_.subscription_id()},
+         {"gcp.project_id", subscription_.project_id()}},
+        std::move(links), options);
+    lk.lock();
+    auto lease_span = lease_span_by_nonce_.find(nonce);
+    if (lease_span != lease_span_by_nonce_.end()) {
+      lease_span->second->End();
+    }
```

Comment:
When would this happen?

Reply:
If there was an outstanding modack using that number and it had never returned/ended.

Reply:
So, it sounds then that it would only happen when we've erroneously reused an id.  If so, I would suggest that a better course of action would be to not do that.  You could check during allocation, or (as mentioned elsewhere) use a name/id that is guaranteed to be unique.

Reply:
ack, addressed below